File: engine/combing.py
"""Cross-field combing for QuadForge.

Combs the 4-RoSy cross-field to produce a consistent 2-direction field
(choosing one arm of the cross per face) across the surface. Where
consistency breaks, seam edges are introduced.

Algorithm:
  1. BFS from a seed face across the mesh
  2. At each edge, choose the matching (out of 4 rotations) that best
     aligns the neighbor's field with the current face's combed direction
  3. Edges where no good match exists become seam edges
"""

from __future__ import annotations

import numpy as np
from collections import deque
from typing import Dict, List, Optional, Set, Tuple


def comb_field(
    num_faces: int,
    field_angles: np.ndarray,
    face_frames_e1: np.ndarray,
    face_frames_e2: np.ndarray,
    face_normals: np.ndarray,
    edge_to_faces: Dict[Tuple[int, int], List[int]],
    transport_angles: Dict[Tuple[int, int], float],
    feature_edges: Optional[Set[Tuple[int, int]]] = None,
) -> Tuple[np.ndarray, np.ndarray, Set[Tuple[int, int]]]:
    """Comb the 4-RoSy field into a consistent 2-direction field.

    Parameters
    ----------
    num_faces : int
    field_angles : (F,) field angle per face in local frame
    face_frames_e1, face_frames_e2, face_normals : (F, 3) per-face frames
    edge_to_faces : edge → face list adjacency
    transport_angles : edge → parallel transport angle
    feature_edges : optional set of feature edge keys

    Returns
    -------
    combed_angles : (F,) combed field angles (consistent orientation)
    period_jumps : (E_seam,) integer period jump at each seam edge
    seam_edges : set of edge keys where seams were cut
    """
    feature_edges = feature_edges or set()

    # Build face adjacency (face_i → [(face_j, edge_key), ...])
    face_adj: Dict[int, List[Tuple[int, Tuple[int, int]]]] = {}
    for edge_key, face_list in edge_to_faces.items():
        if len(face_list) != 2:
            continue
        fi, fj = face_list
        face_adj.setdefault(fi, []).append((fj, edge_key))
        face_adj.setdefault(fj, []).append((fi, edge_key))

    combed = np.copy(field_angles)
    visited = np.zeros(num_faces, dtype=bool)
    # Per-face rotation index k ∈ {0,1,2,3}: how many π/2 rotations applied
    rotation_k = np.zeros(num_faces, dtype=np.int32)

    seam_edges: Set[Tuple[int, int]] = set()

    # BFS from face 0
    queue = deque()
    seed = 0
    visited[seed] = True
    queue.append(seed)

    while queue:
        fi = queue.popleft()
        theta_i = combed[fi]

        for fj, edge_key in face_adj.get(fi, []):
            phi = transport_angles.get(edge_key, 0.0)
            theta_j_raw = field_angles[fj]

            if not visited[fj]:
                # Find best k ∈ {0,1,2,3} such that
                # theta_j_raw + k*π/2 best matches theta_i + phi
                target = theta_i + phi
                best_k = 0
                best_diff = float('inf')

                for k in range(4):
                    candidate = theta_j_raw + k * np.pi / 2.0
                    diff = candidate - target
                    # Wrap to [-π, π]
                    diff = (diff + np.pi) % (2 * np.pi) - np.pi
                    if abs(diff) < best_diff:
                        best_diff = abs(diff)
                        best_k = k

                combed[fj] = theta_j_raw + best_k * np.pi / 2.0
                rotation_k[fj] = best_k
                visited[fj] = True
                queue.append(fj)
            else:
                # Already visited — check consistency
                theta_j_combed = combed[fj]
                expected = theta_i + phi
                diff = theta_j_combed - expected
                diff = (diff + np.pi) % (2 * np.pi) - np.pi

                # If not close to a multiple of π/2, this is a seam
                k_jump = round(diff / (np.pi / 2.0))
                residual = abs(diff - k_jump * np.pi / 2.0)

                if residual > 0.1 or abs(k_jump) > 0:
                    seam_edges.add(edge_key)

    # Handle unvisited faces (disconnected components)
    for fi in range(num_faces):
        if not visited[fi]:
            visited[fi] = True

    return combed, rotation_k, seam_edges


def compute_seam_cut(
    num_faces: int,
    seam_edges: Set[Tuple[int, int]],
    singularity_vertices: np.ndarray,
    edge_to_faces: Dict[Tuple[int, int], List[int]],
    vertices: np.ndarray,
    faces: np.ndarray,
) -> Set[Tuple[int, int]]:
    """Compute a minimal seam-cut graph connecting all singularities.

    The cut graph ensures the surface is topologically a disk (for
    genus-0) by connecting singularity vertices through shortest paths
    on the dual graph.

    Parameters
    ----------
    num_faces : int
    seam_edges : initial seam edges from combing
    singularity_vertices : array of singular vertex indices
    edge_to_faces : adjacency
    vertices, faces : mesh data

    Returns
    -------
    final_seams : set of edge keys forming the complete seam cut
    """
    final_seams = set(seam_edges)

    if len(singularity_vertices) < 2:
        return final_seams

    # Build vertex adjacency with edge weights (edge lengths)
    vert_adj: Dict[int, List[Tuple[int, float, Tuple[int, int]]]] = {}
    for edge_key in edge_to_faces:
        v0, v1 = edge_key
        dist = float(np.linalg.norm(vertices[v1] - vertices[v0]))
        vert_adj.setdefault(v0, []).append((v1, dist, edge_key))
        vert_adj.setdefault(v1, []).append((v0, dist, edge_key))

    # Connect singularities via shortest paths (Dijkstra)
    sing_set = set(int(v) for v in singularity_vertices)

    # Simple approach: connect each singularity to the nearest one
    connected = set()
    if len(sing_set) > 0:
        start = next(iter(sing_set))
        connected.add(start)

        while connected != sing_set:
            best_path = None
            best_dist = float('inf')

            for src in connected:
                path, dist = _dijkstra_to_set(
                    src, sing_set - connected, vert_adj, len(vertices)
                )
                if path is not None and dist < best_dist:
                    best_path = path
                    best_dist = dist

            if best_path is None:
                break

            # Add path edges to seams
            for i in range(len(best_path) - 1):
                v0, v1 = best_path[i], best_path[i + 1]
                ek = (min(v0, v1), max(v0, v1))
                final_seams.add(ek)
            connected.add(best_path[-1])

    return final_seams


def _dijkstra_to_set(
    src: int,
    targets: Set[int],
    vert_adj: Dict[int, List[Tuple[int, float, Tuple[int, int]]]],
    num_verts: int,
) -> Tuple[Optional[List[int]], float]:
    """Dijkstra from src to nearest vertex in targets."""
    import heapq

    dist = np.full(num_verts, np.inf, dtype=np.float64)
    prev = np.full(num_verts, -1, dtype=np.int32)
    dist[src] = 0.0
    heap = [(0.0, src)]

    while heap:
        d, u = heapq.heappop(heap)
        if d > dist[u]:
            continue
        if u in targets and u != src:
            # Reconstruct path
            path = []
            v = u
            while v != -1:
                path.append(v)
                v = int(prev[v])
            return list(reversed(path)), d
        for v, w, _ in vert_adj.get(u, []):
            nd = d + w
            if nd < dist[v]:
                dist[v] = nd
                prev[v] = u
                heapq.heappush(heap, (nd, v))

    return None, float('inf')

File: engine/src/param/combing.cpp
/**
 * combing.cpp — Cross-field combing and seam cut computation.
 *
 * Resolves the 4-fold ambiguity of the 4-RoSy field by BFS from a seed face.
 * Seam edges are introduced where the combing is topologically forced to jump.
 */

#include "../../include/quadforge/param/combing.h"
#include "../../include/quadforge/field/singularity.h"
#include "../../include/quadforge/field/cross_field.h"   // parallel_transport_angle

#include <cmath>
#include <queue>
#include <deque>
#include <algorithm>
#include <unordered_map>
#include <unordered_set>

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

namespace qf {

// -----------------------------------------------------------------------
// Nearest multiple of π/2 rounding
// -----------------------------------------------------------------------
static double snap_to_pi2(double x) {
    return std::round(x / (M_PI/2.0)) * (M_PI/2.0);
}

// -----------------------------------------------------------------------
// comb_field
// -----------------------------------------------------------------------

CombingResult comb_field(
    const HalfEdgeMesh& mesh,
    const CrossField&   field,
    const FeatureData&  features)
{
    int nf = mesh.num_faces();
    CombingResult result;
    result.face_angles.assign(nf, 0.0);

    if (nf == 0) return result;

    // BUG FIX (Bug 3): Build feature-edge key set so the BFS can prefer
    // placing seam cuts along feature edges (natural seam positions).
    auto edge_key = [](int a, int b) -> int64_t {
        int lo = std::min(a,b), hi = std::max(a,b);
        return ((int64_t)lo << 32) | (uint32_t)hi;
    };
    std::unordered_set<int64_t> feat_keys;
    for (auto& [lo, hi] : features.hard_edges)
        feat_keys.insert(edge_key(lo, hi));

    // BFS combing — assigns a globally consistent angle to every face.
    // The BFS spanning tree is "combed" consistently by construction;
    // seams only arise on non-tree (back/cross) edges examined below.
    std::vector<bool> visited(nf, false);

    // he_parent[fj] = the half-edge through which fj was first visited.
    // Used to build the set of tree edges so we can identify non-tree edges.
    std::vector<int> he_parent(nf, -1);

    // BUG FIX (B2 — param/combing v110): Two-pass BFS to guarantee feature
    // edges become non-tree (seam) edges whenever a non-feature path exists.
    //
    // Previous approach: used a deque with push_front / push_back to
    // "prioritise" non-feature edges.  The logic was fatally wrong because
    // `visited[fj] = true` and `he_parent[fj] = he` were set BEFORE the
    // push-order decision, so fj was always claimed by the first (feature)
    // edge that discovered it — the deque ordering only affected when fj's
    // OWN neighbours were explored, not which edge claimed fj.  Feature
    // edges thus became tree edges as often as any other edge, defeating the
    // stated intent entirely.
    //
    // Two-pass fix:
    //   Pass 1 — standard BFS, SKIPPING feature edges entirely.  This fills
    //            every face reachable without crossing a feature edge.
    //   Pass 2 — for any face left unvisited after Pass 1 (only reachable
    //            via a feature edge), visit it using all edges including
    //            feature edges.  Each such unvisited face becomes the seed
    //            of a new mini-BFS that again skips feature edges internally,
    //            until no more unvisited faces remain.
    //
    // Result: feature edges are left as non-tree edges (seam candidates)
    // in every case where a non-feature path exists, and are used only as
    // a last resort for topologically isolated face groups.  This matches
    // the Bommes (2013) recommendation that seam cuts prefer natural feature
    // boundaries to minimise parametrization distortion.
    //
    // The multi-component fix (Bug 4) is preserved: after both passes are
    // exhausted for the current component, the outer seed loop below finds
    // the next unvisited face and repeats.

    // Helper: BFS one wave from a seed face, optionally skipping feature edges.
    // Records he_parent and visited.  Returns nothing — modifies captures.
    std::deque<int> q;
    auto run_bfs = [&](int seed_face, bool skip_feature) {
        result.face_angles[seed_face] = field.face_frames[seed_face];
        visited[seed_face] = true;
        q.push_back(seed_face);

        while (!q.empty()) {
            int fi = q.front(); q.pop_front();
            double ai = result.face_angles[fi];

            auto hes = mesh.face_half_edges(fi);
            for (int he : hes) {
                int tw = mesh.half_edge(he).twin;
                if (tw < 0) continue;
                int fj = mesh.half_edge(tw).face;
                if (fj < 0 || visited[fj]) continue;

                int from_v = mesh.half_edge(mesh.half_edge(he).prev).vertex;
                int to_v   = mesh.half_edge(he).vertex;
                bool is_feat = feat_keys.count(edge_key(from_v, to_v)) > 0;

                // Pass 1: skip feature edges so they remain non-tree.
                if (skip_feature && is_feat) continue;

                // Parallel transport from fi to fj and comb fj.
                double phi            = parallel_transport_angle(mesh, he);
                double transported_ai = ai + phi;
                double aj_raw         = field.face_frames[fj];
                double diff           = aj_raw - transported_ai;
                double k_rot          = std::round(diff / (M_PI / 2.0));
                double aj_combed      = aj_raw - k_rot * (M_PI / 2.0);

                result.face_angles[fj] = aj_combed;
                he_parent[fj]          = he;
                visited[fj]            = true;
                q.push_back(fj);
            }
        }
    };

    // ---- Pass 1: no-feature-edge BFS over all components ----
    run_bfs(0, /*skip_feature=*/true);

    for (int seed = 1; seed < nf; ++seed) {
        if (visited[seed]) continue;
        run_bfs(seed, /*skip_feature=*/true);
    }

    // ---- Pass 2: visit any faces still unreached (feature-edge-only access) ----
    // These are faces that are only reachable by crossing a feature edge.
    // We now allow feature edges but still prefer non-feature paths from
    // the seed so that at least one non-feature tree edge roots each group.
    for (int seed = 0; seed < nf; ++seed) {
        if (visited[seed]) continue;
        // Allow all edges (including feature) for this isolated group.
        run_bfs(seed, /*skip_feature=*/false);
    }

    // BUG FIX (Bug 1): Detect seams on NON-TREE edges.
    //
    // After BFS, every face has a combed angle.  For each interior edge
    // that was NOT a BFS tree edge, check whether the combing is consistent
    // across it.  If it is not (i.e. the angle jump is not a multiple of π/2
    // within tolerance), the edge is a seam.  We record the integer period
    // jump (0–3) for each seam edge.
    //
    // Build BFS tree edge set: (lo, hi) pairs.
    std::unordered_set<int64_t> tree_edge_keys;
    tree_edge_keys.reserve(nf);
    for (int fj = 0; fj < nf; ++fj) {
        int he = he_parent[fj];
        if (he < 0) continue;
        int from_v = mesh.half_edge(mesh.half_edge(he).prev).vertex;
        int to_v   = mesh.half_edge(he).vertex;
        tree_edge_keys.insert(edge_key(from_v, to_v));
    }

    int nhe = mesh.num_half_edges();
    for (int he = 0; he < nhe; ++he) {
        int tw = mesh.half_edge(he).twin;
        if (tw < 0 || tw < he) continue;   // boundary or already seen twin

        int fi = mesh.half_edge(he).face;
        int fj = mesh.half_edge(tw).face;
        if (fi < 0 || fj < 0) continue;

        int from_v = mesh.half_edge(mesh.half_edge(he).prev).vertex;
        int to_v   = mesh.half_edge(he).vertex;
        int64_t ek = edge_key(from_v, to_v);

        // Skip tree edges — they are combed consistently by construction.
        if (tree_edge_keys.count(ek)) continue;

        // Compute the expected angle of fj given fi's combed angle and
        // the parallel transport across this edge.
        double phi = parallel_transport_angle(mesh, he);
        double ai  = result.face_angles[fi];
        double aj  = result.face_angles[fj];

        double transported_ai = ai + phi;
        double diff = aj - transported_ai;

        // Round to nearest multiple of π/2 to get the integer period jump.
        double k_real = diff / (M_PI / 2.0);
        int    k_int  = (int)std::round(k_real);

        // Residual after removing the integer rotation: if non-zero this
        // edge is a genuine seam (the combing is inconsistent).
        double residual = std::abs(diff - k_int * (M_PI / 2.0));

        // Tolerance: π/8 (22.5°) — anything larger is a genuine inconsistency.
        if (residual > M_PI / 8.0 || k_int != 0) {
            int lo = std::min(from_v, to_v);
            int hi = std::max(from_v, to_v);
            result.seam_edges.push_back({lo, hi});

            // BUG FIX (Bug 2): record the period jump (0–3) for this seam edge.
            // Normalise k_int to [0, 3].
            int pj = ((k_int % 4) + 4) % 4;
            result.period_jumps.push_back(pj);
        }
    }

    // Deduplicate seam edges (keep first occurrence which also keeps
    // the corresponding period_jump in sync).
    {
        std::unordered_set<int64_t> seen;
        seen.reserve(result.seam_edges.size());
        std::vector<std::pair<int,int>> dedup_seams;
        std::vector<int>               dedup_jumps;
        for (int k = 0; k < (int)result.seam_edges.size(); ++k) {
            auto& [lo, hi] = result.seam_edges[k];
            int64_t ek = edge_key(lo, hi);
            if (seen.insert(ek).second) {
                dedup_seams.push_back(result.seam_edges[k]);
                dedup_jumps.push_back(result.period_jumps[k]);
            }
        }
        result.seam_edges   = std::move(dedup_seams);
        result.period_jumps = std::move(dedup_jumps);
    }

    return result;
}

// -----------------------------------------------------------------------
// Minimal seam cut via shortest-path spanning tree
// -----------------------------------------------------------------------

EdgeSet compute_minimal_seam_cut(
    const HalfEdgeMesh&                    mesh,
    const std::vector<SingularityInfo>&    singularities)
{
    EdgeSet seam_cut;
    if (singularities.size() < 2) return seam_cut;

    int nv = mesh.num_vertices();
    // Build adjacency list with edge lengths as weights
    std::vector<std::vector<std::pair<int,double>>> adj(nv);
    int nhe = mesh.num_half_edges();
    for (int he = 0; he < nhe; ++he) {
        int tw = mesh.half_edge(he).twin;
        if (tw < 0 || tw < he) continue;
        int from = mesh.half_edge(mesh.half_edge(he).prev).vertex;
        int to   = mesh.half_edge(he).vertex;
        double len = (mesh.vertex_pos(to) - mesh.vertex_pos(from)).norm();
        adj[from].push_back({to, len});
        adj[to].push_back({from, len});
    }

    // Connect each singularity to the first one via Dijkstra
    int root = singularities[0].vertex_index;
    for (size_t i = 1; i < singularities.size(); ++i) {
        int dst = singularities[i].vertex_index;

        // Dijkstra
        const double INF = 1e30;
        std::vector<double> dist(nv, INF);
        std::vector<int>    prev(nv, -1);
        dist[root] = 0;
        using PD = std::pair<double,int>;
        std::priority_queue<PD, std::vector<PD>, std::greater<PD>> pq;
        pq.push({0.0, root});
        while (!pq.empty()) {
            auto [d, u] = pq.top(); pq.pop();
            if (d > dist[u]) continue;
            if (u == dst) break;
            for (auto& [v, w] : adj[u]) {
                if (dist[u]+w < dist[v]) {
                    dist[v] = dist[u]+w;
                    prev[v] = u;
                    pq.push({dist[v], v});
                }
            }
        }

        // Trace path dst → root
        for (int v = dst; v != root && v != -1; v = prev[v]) {
            int p = prev[v];
            if (p < 0) break;
            int lo = std::min(v,p), hi = std::max(v,p);
            seam_cut.push_back({lo, hi});
        }
    }

    std::sort(seam_cut.begin(), seam_cut.end());
    seam_cut.erase(std::unique(seam_cut.begin(), seam_cut.end()), seam_cut.end());
    return seam_cut;
}

} // namespace qf
